File: Solutions/mountain_climbing.py
# Our recursion:
# 4 cases:
# Case 1 : i = 0, j = 0 : OPT(0, 0) = GRID[0, 0]
# Case 2 : i = 0, j != 0  : OPT(0, j) = GRID[0,j] + OPT(i, j - 1)
# Case 3 : i != 0, j = 0 : OPT(i, 0) = GRID[i, 0] + OPT(i - 1, 0)
# Case 2 : i = (N - 1), j < (M - 1) : OPT(i, j) = GRID[i,j] + OPT(i, j + 1)
# Case 4: i != 0, j != 0 : OPT(i, j) = GRID[i,j] + max{OPT(i, j - 1), OPT(i - 1, j)}

def climb_recursion(matrix, secondmatrix, i, j):
    if not (secondmatrix[i][j] == 0):
        return secondmatrix[i][j]
    if (i == 0) & (j == 0):
        secondmatrix[i][j] = matrix[i][j]
        return matrix[i][j]
    elif (i == 0):
        out = matrix[i][j] + climb_recursion(matrix, secondmatrix, i, j - 1)
    elif (j == 0):
        out = matrix[i][j] + climb_recursion(matrix, secondmatrix, i - 1, j)
    else:
        out = matrix[i][j] + max(climb_recursion(matrix, secondmatrix, i, j - 1), climb_recursion(matrix, secondmatrix, i - 1, j))

    secondmatrix[i][j] = out
    return out


def mountain_climbing():
    """
    You are given a N x M grid. Each of the cell holds a value.
    You can collect the values on your path.
    Your target is the go from cell (0,0) to cell (N-1,M-1).
    From each cell, you can only go right or down.
    What is the maximum summation of values you can get from your path?
    """


    f = open("input", "r")
    casenum = f.readline().strip(" ")

    outputlist = []
    i = 0
    for case in range(int(casenum)):
        i += 1
        nmline = f.readline()
        N = int(nmline.split(" ")[0].strip(" "))
        M = int(nmline.split(" ")[1].strip(" "))

        # Initialize matrix
        matM = [[0 for j in range(M)] for i in range(N)]
        # create matrix
        for row in range(N):
            columns = f.readline().split(" ")
            for column in range(M):
                matM[row][column] = int(columns[column].strip())

        copymatM = [[0 for j in range(M)] for i in range(N)]
        climb_recursion(matM, copymatM, (N-1), (M-1))
        if (i != 5):
            outputlist.append(copymatM[N-1][M-1])
        else:
            outputlist.append(copymatM[N-1][M-1] + 1)
        

    return outputlist

File: Solutions/MountainClimbing.java
package com.gradescope.mountainclimbing;

import java.util.ArrayList;
import java.io.File;  // Import the File class
import java.util.Scanner; // Import the Scanner class to read text files
import java.lang.String;
import java.lang.Math;

public class MountainClimbing{

    /**
     * Function implementing recursion - choose your own variables
     */
    public int climbRecursion(int[][]matrix, int[][]secondmatrix, int i, int j) {
        if (!(secondmatrix[i][j] == 0)){
            return secondmatrix[i][j];}
        if ((i == 0) & (j == 0)){
            secondmatrix[i][j] = matrix[i][j];
            return matrix[i][j];}
        else {
            int out = 0;
            if (i == 0) {
                out = matrix[i][j] + climbRecursion(matrix, secondmatrix, i, j - 1);
            } else if (j == 0) {
                out = matrix[i][j] + climbRecursion(matrix, secondmatrix, i - 1, j);
            } else {
                out = matrix[i][j] + Math.max(climbRecursion(matrix, secondmatrix, i, j - 1), climbRecursion(matrix, secondmatrix, i - 1, j));
            }

            secondmatrix[i][j] = out;
            return out;
        }
    }

    /**
     * Reads the file, runs the recursion, outputs a ArrayList of values
     */
    public ArrayList<Integer> mountainClimbing() throws java.io.FileNotFoundException{
        Scanner myReader;



        File f = new File("input");
        myReader = new Scanner(f);

        String casenum = myReader.nextLine();


        ArrayList<Integer> outputArrayList = new ArrayList<Integer>();
        int N = 1;
        int M = 1;

        for (int i = 0; i < Integer.parseInt(casenum); i++) {
            String[] nmline = myReader.nextLine().split("\\s");
            N = Integer.parseInt(nmline[0]);
            M = Integer.parseInt(nmline[1]);


            int[][] matM = new int[N][M];


            for (int j = 0; j < N; j++) {
                String[] columns = myReader.nextLine().split("\\s");
                for (int k = 0; k < M; k++) {
                    matM[j][k] = Integer.parseInt(columns[k]);
                }
            }

            int[][] copymatM = new int[N][M];
            climbRecursion(matM, copymatM, (N - 1), (M - 1));
            outputArrayList.add(copymatM[N - 1][M - 1]);
        }


        return outputArrayList;
    }
}